provider/openstack/provider.go

```diff
@@ -1047,22 +1047,30 @@ func (e *Environ) getKeystoneDataSource(mu *sync.Mutex, datasource *simplestream
 		return *datasource, nil
 	}
 
-	client := e.client()
-	if !client.IsAuthenticated() {
-		if err := authenticateClient(client); err != nil {
+	cl := e.client()
+	if !cl.IsAuthenticated() {
+		if err := authenticateClient(cl); err != nil {
 			return nil, err
 		}
 	}
 
-	url, err := makeServiceURL(client, keystoneName, "", nil)
+	serviceURL, err := makeServiceURL(cl, keystoneName, "", nil)
 	if err != nil {
 		return nil, errors.NewNotSupported(err, fmt.Sprintf("cannot make service URL: %v", err))
 	}
 	verify := utils.VerifySSLHostnames
 	if !e.Config().SSLHostnameVerification() {
 		verify = utils.NoVerifySSLHostnames
 	}
-	*datasource = simplestreams.NewURLDataSource("keystone catalog", url, verify, simplestreams.SPECIFIC_CLOUD_DATA, false)
+	cfg := simplestreams.Config{
+		Description:          "keystone catalog",
+		BaseURL:              serviceURL,
+		HostnameVerification: verify,
+		Priority:             simplestreams.SPECIFIC_CLOUD_DATA}
+	if err := cfg.Validate(); err != nil {
+		return nil, errors.Annotate(err, "simplestreams config validation failed")
+	}
+	*datasource = simplestreams.NewDataSource(cfg)
```

Comment:
Overwriting of pointers passed as input is quite unusual pattern for Go code. Any ideas why the original method was implemented like this?

Reply:
Agreed.  Best guess: looks like it was an attempt to ensure that the provider only got a NewDataSource once and reused it.  It may be easy to remove while looking at the "accept an interface return a structure" changes I'm attempting now.